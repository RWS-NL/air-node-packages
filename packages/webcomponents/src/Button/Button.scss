@import '../styles/vars';

.button {
  @extend %fontNormal;
  background-color: $skyBlue100;
  border-radius: 0;
  float: right;
  height: 3.2rem;
  line-height: initial;
  margin-right: 1%;
  text-transform: none;

  &Label {
    @extend %fontNormal;
    color: $white;
    font-size: 1rem;
    text-transform: none;
  }

  &Secondary {
    background-color: $yellow100;
    color: $textBlack;

    > span {
      color: $textBlack;
    }
  }

  &Outlined {
    background-color: $white;
    color: $skyBlue100;

    > span {
      color: $skyBlue100;
    }

    &Secondary {
      background-color: $white;
      color: $textBlack;

      > span {
        color: $textBlack;
      }
    }
  }

  &Disabled {
    background-color: $grey35 !important;
    border-color: $grey35 !important;
    box-shadow: none !important;
  }
}
