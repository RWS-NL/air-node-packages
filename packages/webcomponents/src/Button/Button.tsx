import MUIButton, { ButtonProps as MUIButtonProps } from '@material-ui/core/Button';
import css from './Button.scss';
import classnames from 'classnames';
import React, { FC, ReactNode, memo } from 'react';
import { makeStyles, Theme, createStyles } from '@material-ui/core';

export interface ButtonProps {
  /** The variant of the button */
  variant: 'text' | 'outlined' | 'contained' | undefined;
  /** The color type of the button */
  color: 'primary' | 'secondary' | 'default' | 'inherit' | undefined;
  /** The label to be displayed inside the button */
  label: ReactNode;
  /** Whether this button should be disabled */
  disabled?: boolean;
  /** Data-qa tag to apply to the search bar and input element */
  'data-qa'?: string;
  /** Custom CSS classes to pass to the button */
  customclasses?: string | string[];
  /** Custom CSS classes to pass to the button label */
  customlabelclasses?: string | string[];
  /** The action that should be triggered when clicking the button */
  onClick(): unknown;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({ buttonShadow: { boxShadow: theme.shadows[1], border: 'transparant' } })
);

/**
 * Constructs a button using pre-defined Rijkswaterstaat styling
 * @param props Props to pass to the button
 * @example
 * <Button label='button' onClick={() => undefined} variant='contained' color='primary' />
 */
export const Button: FC<MUIButtonProps<'button', ButtonProps>> = props => {
  const classes = useStyles();

  return (
    <MUIButton
      {...props}
      data-qa={props['data-qa']}
      onClick={props.onClick}
      variant={props.variant}
      color={props.color}
      disabled={props.disabled}
      className={classnames(props.customclasses)}
      classes={{
        root: css.button,
        label: classnames(css.buttonLabel, props.customlabelclasses),
        disabled: css.buttonDisabled,
        containedPrimary: classnames(css.buttonPrimary, classes.buttonShadow),
        containedSecondary: classnames(css.buttonSecondary, classes.buttonShadow),
        outlined: classnames(css.buttonOutlined, classes.buttonShadow),
        outlinedPrimary: classnames(css.buttonOutlined, classes.buttonShadow),
        outlinedSecondary: classnames(css.buttonOutlinedSecondary, classes.buttonShadow)
      }}
      href={undefined}
    >
      {props.label}
    </MUIButton>
  );
};

export default memo(Button);
