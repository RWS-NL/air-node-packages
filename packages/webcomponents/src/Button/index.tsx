import { createStyles, makeStyles, Theme } from '@material-ui/core';
import MUIButton, { ButtonProps as MUIButtonProps } from '@material-ui/core/Button';
import classnames from 'classnames';
import React, { memo, ReactNode } from 'react';
import css from './Button.scss';

export interface ButtonProps {
  /** The variant of the button */
  variant: 'text' | 'outlined' | 'contained' | undefined;
  /** The color type of the button */
  color: 'primary' | 'secondary' | 'default' | 'inherit' | undefined;
  /** The label to be displayed inside the button */
  label: ReactNode;
  /** Whether this button should be disabled */
  disabled?: boolean;
  /** Data-qa tag to apply to the search bar and input element */
  'data-qa'?: string;
  /** Custom CSS classes to pass to the button */
  customclasses?: string | string[];
  /** Custom CSS classes to pass to the button label */
  customlabelclasses?: string | string[];
  /** The action that should be triggered when clicking the button */
  onClick(): unknown;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({ buttonShadow: { boxShadow: theme.shadows[1], border: 'transparant' } })
);

/**
 * Constructs a button using pre-defined Rijkswaterstaat styling
 * @param props Props to pass to the button
 * @example
 * ```jsx
 * <Button label='button' onClick={() => undefined} variant='contained' color='primary' />
 * ```
 */
export const Button = memo((props: MUIButtonProps<'button', ButtonProps>) => {
  const classes = useStyles();

  return (
    <MUIButton
      {...props}
      data-qa={props['data-qa']}
      onClick={props.onClick}
      variant={props.variant}
      color={props.color}
      disabled={props.disabled}
      className={classnames(props.customclasses)}
      classes={{
        root: css.button,
        label: classnames(css.buttonLabel, props.customlabelclasses),
        disabled: css.buttonDisabled,
        containedPrimary: classnames(css.buttonPrimary, classes.buttonShadow),
        containedSecondary: classnames(css.buttonSecondary, classes.buttonShadow),
        outlined: classnames(css.buttonOutlined, classes.buttonShadow),
        outlinedPrimary: classnames(css.buttonOutlined, classes.buttonShadow),
        outlinedSecondary: classnames(css.buttonOutlinedSecondary, classes.buttonShadow)
      }}
      href={undefined}
    >
      {props.label}
    </MUIButton>
  );
});

export default Button;
