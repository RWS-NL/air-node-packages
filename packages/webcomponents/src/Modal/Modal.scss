@import '../styles/vars';

%modalButton {
  width: 11.5rem;
}

.modal {
  > div > div {
    border-radius: 0;
    max-width: unset;
    min-width: 45vw;
  }
}

.content {
  padding: 0 3rem 2rem;
}

.title {
  padding: 3rem 3rem 1rem;

  > * {
    color: $logoBlue;
    font-size: 1.1rem;
    font-weight: bolder;
  }
}

.closeIcon {
  bottom: 3rem;
  left: 2.7rem;
  position: relative;

  &SVG {
    height: 0.7em;
    width: 0.7em;
  }
}
