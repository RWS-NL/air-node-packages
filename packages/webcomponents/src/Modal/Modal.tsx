import Dialog from '@material-ui/core/Dialog';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import { objectHasProperty } from '@rws-air/utils';
import React, { FC, memo, ReactNode } from 'react';
import { dataQa } from '../constants';
import css from './Modal.scss';

export interface ModalQAs {
  /** Data-qa applied to the main modal */
  modal?: dataQa;
  /** Data-qa applied to the title */
  title?: dataQa;
  /** Data-qa applied to the content */
  content?: dataQa;
}

export interface ModalProps {
  /** Whether the modal should be opened or not */
  open: boolean;
  /** Object of data-qa tags to pass to the modal */
  modalqas?: ModalQAs;
  /** Title for the modal */
  topic: string | ReactNode;
  /** Whether clicking on the backdrop should close the modal (triggers closeAction)  */
  disableBackdropClick?: boolean;
  /** Whether pressing the escape key should close the modal (triggers closeAction) */
  disableEscapeKeyDown?: boolean;
}

/** Creates a modal using pre-defined Rijkswaterstaat styling */
export const Modal: FC<ModalProps> = props => {
  return (
    <Dialog
      disableBackdropClick={objectHasProperty(props, 'disableBackdropClick') ? props.disableBackdropClick : true}
      disableEscapeKeyDown={objectHasProperty(props, 'disableEscapeKeyDown') ? props.disableEscapeKeyDown : true}
      open={Boolean(props.open)}
      classes={{ root: css.modal }}
      data-qa={props.modalqas?.modal || 'modal'}
    >
      <DialogTitle className={css.title} data-qa={props.modalqas?.title || 'modal-title'}>
        {props.topic}
      </DialogTitle>
      <DialogContent data-qa={props.modalqas?.content || 'modal-content'} >
        {props.children}
      </DialogContent>
    </Dialog>
  );
};

export default memo(Modal);