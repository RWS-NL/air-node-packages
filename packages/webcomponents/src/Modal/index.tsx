import Dialog, { DialogProps } from '@mui/material/Dialog';
import DialogContent, { DialogContentProps } from '@mui/material/DialogContent';
import DialogTitle, { DialogTitleProps } from '@mui/material/DialogTitle';
import Grid from '@mui/material/Grid';
import IconButton from '@mui/material/IconButton';
import CloseIcon from '@mui/icons-material/Close';
import React, { memo, ReactNode } from 'react';
import css from './Modal.scss';

export interface ModalQAs {
  /** Data-qa applied to the main modal */
  modal?: string;
  /** Data-qa applied to the title */
  title?: string;
  /** Data-qa applied to the content */
  content?: string;
}

export interface ModalProps {
  /** Whether the modal should be opened or not */
  open: boolean;
  /** The content to show in the modal */
  dialogContent: JSX.Element;
  /** Title for the modal */
  topic: string | ReactNode;
  /** Any othe props passed to the Dialog */
  DialogProps?: DialogProps;
  /** Any other props passed to the DialogContent */
  DialogContentProps?: DialogContentProps;
  /** Any other props passed to the DialogTitle */
  DialogTitleProps?: DialogTitleProps;
  /** Object of data-qa tags to pass to the modal */
  modalqas?: ModalQAs;

  /** The action to trigger to close this modal */
  closeAction(): void;
}

/**
 * Constructs a modal using pre-defined Rijkswaterstaat styling
 * @param props Props to pass to the modal - {@link ModalProps}
 * @example
 * ```jsx
 * <Modal
 *   topic='topic'
 *   modalqas={{
 *     modal: 'modal',
 *     content: 'content',
 *     title: 'title',
 *   }}
 *   dialogContent={<div>SAMPLE</div>}
 *   open={true}
 * />
 * ```
 */
export const Modal = memo(({ DialogContentProps, DialogTitleProps, DialogProps, ...props }: ModalProps) => {
  return (
    <Dialog
      {...DialogProps}
      disableEscapeKeyDown
      open={Boolean(props.open)}
      classes={{ root: css.modal }}
      data-qa={props.modalqas?.modal || 'modal'}
      onClose={(_event, reason) => {
        if (reason !== 'backdropClick') {
          props.closeAction();
        }
      }}
    >
      <DialogTitle {...DialogTitleProps} className={css.title} data-qa={props.modalqas?.title || 'modal-title'}>
        <Grid container alignContent='flex-end' alignItems='flex-end' justifyContent='space-between'>
          <Grid item>{props.topic}</Grid>
          <Grid item classes={{ root: css.closeIcon }}>
            <IconButton
              onClick={props.closeAction}
              size='small'
              edge='end'
              disableFocusRipple
              disableTouchRipple
              disableRipple
              color='primary'
            >
              <CloseIcon classes={{ root: css.closeIconSVG }} />
            </IconButton>
          </Grid>
        </Grid>
      </DialogTitle>
      <DialogContent
        {...DialogContentProps}
        className={css.content}
        data-qa={props.modalqas?.content || 'modal-content'}
      >
        {props.dialogContent}
      </DialogContent>
    </Dialog>
  );
});
