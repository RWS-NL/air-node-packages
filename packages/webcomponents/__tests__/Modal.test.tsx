import { IconButton } from '@material-ui/core';
import { Button } from '@src/Button';
import { Modal, ModalProps } from '@src/Modal';
import { shallow, ShallowWrapper } from 'enzyme';
import React, { Fragment } from 'react';

let modal: ShallowWrapper<ModalProps, any>;
const mockCloseAction = jest.fn();

const setup = (isOpen = false, props?: Partial<ModalProps>, DialogContent: () => JSX.Element = () => <Fragment />) =>
  (modal = shallow<ModalProps>(
    <Modal
      topic='Awesome Topic'
      open={isOpen}
      closeAction={mockCloseAction}
      dialogContent={<DialogContent />}
      modalqas={{
        content: 'modal-content',
        modal: 'modal',
        title: 'modal-title'
      }}
      {...props}
    />
  ));

describe('Render Testing', () => {
  test('should include mandatory props', () => {
    setup();
    expect(modal.find(Button)).toHaveLength(0);
    expect(modal.prop('disableBackdropClick')).toBe(true);
    expect(modal.prop('disableEscapeKeyDown')).toBe(true);
  });

  test("should pass all data-qa's", () => {
    setup(true);

    expect(modal.find('[data-qa="modal"]').exists()).toBe(true);
    expect(modal.find('[data-qa="modal-title"]').exists()).toBe(true);
  });
});

describe('Content Testing', () => {
  beforeAll(() => setup(true, {}, () => <div>Content</div>));
  afterAll(() => {
    jest.clearAllMocks();
  });

  test('should contain content', () => {
    expect(modal.find('[data-qa="modal-content"]').children().exists()).toBe(true);
    expect(modal.find('[data-qa="modal-content"]').children()).toMatchSnapshot();
  });

  test('should close when clicking X icon', () => {
    const closeButton = modal.find(IconButton);

    closeButton.simulate('click');

    expect(mockCloseAction).toHaveBeenCalledWith();
    expect(mockCloseAction).toHaveBeenCalledTimes(1);
  });
});

describe('Snapshot Testing', () => {
  test('Required Props', () => {
    setup();
    expect(modal).toMatchSnapshot();
  });
});
