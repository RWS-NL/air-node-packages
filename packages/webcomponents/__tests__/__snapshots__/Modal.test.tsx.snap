// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content Testing should contain content 1`] = `<Component />`;

exports[`Snapshot Testing Required Props 1`] = `
<WithStyles(ForwardRef(Dialog))
  classes={
    Object {
      "root": "modal",
    }
  }
  data-qa="modal"
  disableBackdropClick={true}
  disableEscapeKeyDown={true}
  open={false}
>
  <WithStyles(ForwardRef(DialogTitle))
    className="title"
    data-qa="modal-title"
  >
    <WithStyles(ForwardRef(Grid))
      alignContent="flex-end"
      alignItems="flex-end"
      container={true}
      justify="space-between"
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
      >
        Awesome Topic
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        classes={
          Object {
            "root": "closeIcon",
          }
        }
        item={true}
      >
        <WithStyles(ForwardRef(IconButton))
          color="primary"
          disableFocusRipple={true}
          disableRipple={true}
          disableTouchRipple={true}
          edge="end"
          onClick={[MockFunction]}
          size="small"
        >
          <CloseIcon
            classes={
              Object {
                "root": "closeIconSVG",
              }
            }
          />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))
    className="content"
    data-qa="modal-content"
  >
    <Component />
  </WithStyles(ForwardRef(DialogContent))>
</WithStyles(ForwardRef(Dialog))>
`;
